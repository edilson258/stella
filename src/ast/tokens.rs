#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::utils::location::Location;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Token {
  pub kind: TokenKind,
  pub location: Location,
}

impl Token {
  pub fn new(kind: TokenKind, location: Location) -> Token {
    Token { kind, location }
  }

  pub fn new_end_of_file(location: Location) -> Token {
    Token::new(TokenKind::EOF, location)
  }

  pub fn new_identifier(location: Location, identifier: String) -> Token {
    Token::new(TokenKind::Identifier(identifier), location)
  }

  pub fn new_number(location: Location, number: f64) -> Token {
    Token::new(TokenKind::Number(number), location)
  }

  pub fn new_string(location: Location, string: String) -> Token {
    Token::new(TokenKind::String(string), location)
  }

  pub fn new_keyword_or_identifier(location: Location, keyword: String) -> Token {
    match keyword.as_str() {
      "function" => Token::new(TokenKind::Function, location),
      "local" => Token::new(TokenKind::Local, location),
      "if" => Token::new(TokenKind::If, location),
      "then" => Token::new(TokenKind::Then, location),
      "else" => Token::new(TokenKind::Else, location),
      "elseif" => Token::new(TokenKind::ElseIf, location),
      "end" => Token::new(TokenKind::End, location),
      "while" => Token::new(TokenKind::While, location),
      "do" => Token::new(TokenKind::Do, location),
      "for" => Token::new(TokenKind::For, location),
      "in" => Token::new(TokenKind::In, location),
      "repeat" => Token::new(TokenKind::Repeat, location),
      "until" => Token::new(TokenKind::Until, location),
      "return" => Token::new(TokenKind::Return, location),
      "break" => Token::new(TokenKind::Break, location),
      "true" => Token::new(TokenKind::True, location),
      "false" => Token::new(TokenKind::False, location),
      "nil" => Token::new(TokenKind::Nil, location),
      "type" => Token::new(TokenKind::Type, location),
      "enum" => Token::new(TokenKind::Enum, location),
      _ => Token::new(TokenKind::Identifier(keyword), location),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum TokenKind {
  // keywords
  Function,
  Local,
  If,
  Then,
  Else,
  ElseIf,
  End,
  While,
  Do,
  For,
  In,
  Repeat,
  Until,
  Return,
  Break,
  True,
  False,
  Nil,
  Type,
  Enum,

  // Operators and Delimiters
  Tilde,        // ~
  Plus,         // +
  Minus,        // -
  Star,         // *
  StarAssign,   // *=
  Slash,        // /
  SlashAssign,  // /=
  Percent,      // %
  Equal,        // ==
  NotEqual,     // ~=
  Less,         // <
  Greater,      // >
  LessEqual,    // <=
  GreaterEqual, // >=
  Assign,       // =
  PlusAssign,   // +=
  MinusAssign,  // -=
  And,          // and
  Or,           // or
  Not,          // not
  Hash,         // #
  Comma,        // ,
  Semicolon,    // ;
  Colon,        // :
  DoubleColon,  // ::
  Dot,          // .
  DoubleDot,    // ..
  TripleDot,    // ...
  LeftParen,    // (
  RightParen,   // )
  LeftBrace,    // {
  RightBrace,   // }
  LeftBracket,  // [
  RightBracket, // ]

  // Literals
  Identifier(String),
  Number(f64),
  String(String),

  // Outros
  EOF, // End of file
}
