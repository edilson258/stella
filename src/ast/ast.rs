#![allow(dead_code)]
use super::tokens::Token;
use crate::utils::location::Location;
#[allow(dead_code)]
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Program {
  pub statements: Vec<Statement>,
}

impl Program {
  pub fn new() -> Program {
    Program { statements: vec![] }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Statement {
  AssignStatement(AssignStatement),
  FunctionStatement(FunctionStatement),
  ReturnStatement(ReturnStatement),
  IfStatement(IfStatement),
  WhileStatement(WhileStatement),
  RepeatStatement(RepeatStatement),
  ForStatement(ForStatement),
  BreakStatement(BreakStatement),
  ContinueStatement(ContinueStatement),
  BlockStatement(BlockStatement),
  EmptyStatement(EmptyStatement),
  LocalStatement(LocalStatement),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AssignStatement {
  pub name: String,
  pub value: ExpressionStatement,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FunctionStatement {
  pub name: Token,
  pub arguments: Vec<(Token, Type)>,
  pub return_type: Type,
  pub body: Box<Statement>,
  pub location: Location,
}

impl FunctionStatement {
  pub fn new(
    name: Token,
    arguments: Vec<(Token, Type)>,
    return_type: Type,
    body: Statement,
    location: Location,
  ) -> Self {
    FunctionStatement { name, arguments, return_type, body: Box::new(body), location }
  }
}
#[derive(Debug, Serialize, Deserialize)]
pub struct ReturnStatement {
  pub value: ExpressionStatement,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct IfStatement {
  pub condition: ExpressionStatement,
  pub body: Vec<Statement>,
  pub else_body: Option<Vec<Statement>>,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WhileStatement {
  pub condition: ExpressionStatement,
  pub body: Vec<Statement>,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RepeatStatement {
  pub body: Vec<Statement>,
  pub condition: ExpressionStatement,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ForStatement {
  pub initializer: Option<ExpressionStatement>,
  pub condition: Option<ExpressionStatement>,
  pub increment: Option<ExpressionStatement>,
  pub body: Vec<Statement>,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BreakStatement {
  pub label: Option<String>,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ContinueStatement {
  pub label: Option<String>,
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BlockStatement {
  pub body: Vec<Statement>,
  pub location: Location,
}
impl BlockStatement {
  pub fn new(body: Vec<Statement>, location: Location) -> Self {
    BlockStatement { body, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct EmptyStatement {
  pub location: Location,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LocalStatement {
  pub name: Token,
  pub type_: Option<Type>,
  pub init: ExpressionStatement,
  pub location: Location,
}

impl LocalStatement {
  pub fn new(name: Token, type_: Option<Type>, init: ExpressionStatement, location: Location) -> Self {
    LocalStatement { name, type_, init, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum ExpressionStatement {
  LiteralExpression(LiteralExpression),
  CallExpression(CallExpression),
  UnaryExpression(UnaryExpression),
  BinaryExpression(BinaryExpression),
  DeclarationExpression(DeclarationExpression),
}

impl ExpressionStatement {
  pub fn new_number_literal(value: String, location: Location) -> Self {
    let literal = LiteralExpression::new_number_literal(value, location);
    return ExpressionStatement::LiteralExpression(literal);
  }

  pub fn new_string_literal(value: String, location: Location) -> Self {
    let literal = LiteralExpression::new_string_literal(value, location);
    return ExpressionStatement::LiteralExpression(literal);
  }

  pub fn new_bool_literal(value: bool, location: Location) -> Self {
    let literal = LiteralExpression::new_bool_literal(value, location);
    return ExpressionStatement::LiteralExpression(literal);
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeclarationExpression {
  pub name: Token,
  pub value: Box<ExpressionStatement>,
  pub local: bool,
  pub location: Location,
}

impl DeclarationExpression {
  pub fn new(name: Token, value: Box<ExpressionStatement>, local: bool, location: Location) -> Self {
    DeclarationExpression { name, value, local, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CallExpression {
  pub name: Token,
  pub args: Vec<ExpressionStatement>,
  pub location: Location,
}

impl CallExpression {
  pub fn new(name: Token, args: Vec<ExpressionStatement>, location: Location) -> Self {
    CallExpression { name, args, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UnaryExpression {
  pub operator: UnaryOperator,
  pub operand: Box<ExpressionStatement>,
  pub location: Location,
}

impl UnaryExpression {
  pub fn new(operator: UnaryOperator, operand: Box<ExpressionStatement>, location: Location) -> Self {
    UnaryExpression { operator, operand, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BinaryExpression {
  pub operator: BinaryOperator,
  pub left: Box<ExpressionStatement>,
  pub right: Box<ExpressionStatement>,
  pub location: Location,
}

impl BinaryExpression {
  pub fn new(
    operator: BinaryOperator,
    left: Box<ExpressionStatement>,
    right: Box<ExpressionStatement>,
    location: Location,
  ) -> Self {
    BinaryExpression { operator, left, right, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum LiteralExpression {
  NumberLiteral(NumberLiteral),
  StringLiteral(StringLiteral),
  BoolLiteral(BoolLiteral),
}

impl LiteralExpression {
  pub fn new_number_literal(value: String, location: Location) -> Self {
    LiteralExpression::NumberLiteral(NumberLiteral::new(value, location))
  }

  pub fn new_string_literal(value: String, location: Location) -> Self {
    LiteralExpression::StringLiteral(StringLiteral::new(value, location))
  }

  pub fn new_bool_literal(value: bool, location: Location) -> Self {
    LiteralExpression::BoolLiteral(BoolLiteral::new(value, location))
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NumberLiteral {
  pub value: String,
  pub location: Location,
}

impl NumberLiteral {
  pub fn new(value: String, location: Location) -> Self {
    NumberLiteral { value, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StringLiteral {
  pub value: String,
  pub location: Location,
}

impl StringLiteral {
  pub fn new(value: String, location: Location) -> Self {
    StringLiteral { value, location }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BoolLiteral {
  pub value: bool,
  pub location: Location,
}

impl BoolLiteral {
  pub fn new(value: bool, location: Location) -> Self {
    BoolLiteral { value, location }
  }
}

// Definição de Tipos

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Type {
  Void,
  Number,
  String,
  Bool,
  Identifier(String),
}

impl Type {
  pub fn identifier(name: String) -> Self {
    Type::Identifier(name)
  }

  pub fn new_type(text: String) -> Self {
    match text.as_str() {
      "void" => Type::Void,
      "number" => Type::Number,
      "string" => Type::String,
      "bool" => Type::Bool,
      _ => Type::Identifier(text),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum UnaryOperator {
  Negate,
  Not,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum BinaryOperator {
  Add,
  Subtract,
  Multiply,
  Divide,
  Modulus,
  And,
  Or,
  Equal,
  NotEqual,
  LessThan,
  GreaterThan,
  LessThanOrEqual,
  GreaterThanOrEqual,
}
