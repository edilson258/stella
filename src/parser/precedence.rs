use crate::ast::tokens::TokenKind;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Precedence {
  Assignment,
  Or,
  And,
  Equality,
  Comparison,
  Term,
  Factor,
  Unary,
  Primary,
  Concat,
}

impl Precedence {
  pub fn operators(self) -> &'static [TokenKind] {
    match self {
      Precedence::Or => &[TokenKind::Or],
      Precedence::And => &[TokenKind::And],
      Precedence::Equality => &[TokenKind::Equal, TokenKind::NotEqual],
      Precedence::Comparison => &[TokenKind::Less, TokenKind::LessEqual, TokenKind::Greater, TokenKind::GreaterEqual],
      Precedence::Term => &[TokenKind::Plus, TokenKind::Minus],
      Precedence::Concat => &[TokenKind::DoubleDot],
      Precedence::Factor => &[TokenKind::Star, TokenKind::Slash, TokenKind::Percent],
      _ => &[], // no binary operators for unary and primary precedence :(
    }
  }

  pub fn next(self) -> Precedence {
    match self {
      Precedence::Assignment => Precedence::Or,
      Precedence::Or => Precedence::And,
      Precedence::And => Precedence::Equality,
      Precedence::Equality => Precedence::Comparison,
      Precedence::Comparison => Precedence::Concat,
      Precedence::Concat => Precedence::Term,
      Precedence::Term => Precedence::Factor,
      Precedence::Factor => Precedence::Unary,
      Precedence::Unary => Precedence::Primary,
      Precedence::Primary => Precedence::Primary,
    }
  }
}
