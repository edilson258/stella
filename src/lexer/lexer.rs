use crate::ast::tokens::{Token, TokenKind};
use crate::diagnostics::report::report_and_exit;
use crate::utils::match_number;
use crate::utils::range::Range;

pub struct Lexer<'a> {
  raw: &'a str,
  column: usize,
  line: usize,
  cursor: usize,
  range_start: usize,
  peeked_token: Option<Token>,
  pub file_name: &'a str,
}

impl<'a> Lexer<'a> {
  pub fn new(raw: &'a str, file_name: &'a str) -> Lexer<'a> {
    Lexer { raw, column: 0, range_start: 0, line: 1, cursor: 0, peeked_token: None, file_name }
  }

  pub fn peek_token(&mut self) -> Token {
    if self.peeked_token.is_none() {
      self.peeked_token = Some(self.read_next_token());
    }
    self.peeked_token.clone().unwrap()
  }

  pub fn next_token(&mut self) -> Token {
    if let Some(token) = self.peeked_token.take() {
      return token;
    }
    self.read_next_token()
  }

  fn read_next_token(&mut self) -> Token {
    self.skip_whitespace();
    self.update_current_range();
    if self.is_end() {
      return Token::new(TokenKind::EOF, self.create_range());
    }

    let current_char = self.peek_one();
    match current_char {
      '+' => self.read_check_ahead("+=", TokenKind::Plus, TokenKind::PlusAssign),
      '-' => self.read_comment_or_minus(),
      '*' => self.read_check_ahead("*=", TokenKind::Star, TokenKind::StarAssign),
      '=' => self.read_check_ahead("==", TokenKind::Assign, TokenKind::Equal),
      '~' => self.read_check_ahead("~=", TokenKind::Tilde, TokenKind::NotEqual),
      '<' => self.read_check_ahead("<=", TokenKind::Less, TokenKind::LessEqual),
      '>' => self.read_check_ahead(">=", TokenKind::Greater, TokenKind::GreaterEqual),
      '(' => self.read_simple_token(TokenKind::LeftParen),
      ')' => self.read_simple_token(TokenKind::RightParen),
      '%' => self.read_simple_token(TokenKind::Percent),
      '#' => self.read_simple_token(TokenKind::Hash),
      ',' => self.read_simple_token(TokenKind::Comma),
      ':' => self.read_simple_token(TokenKind::Colon),
      ';' => self.read_simple_token(TokenKind::Semicolon),
      '{' => self.read_simple_token(TokenKind::LeftBrace),
      '}' => self.read_simple_token(TokenKind::RightBrace),
      '[' => self.read_simple_token(TokenKind::LeftBracket),
      ']' => self.read_simple_token(TokenKind::RightBracket),
      '"' => self.read_string(),
      '\'' => self.read_single_quote_string(),
      '/' => self.read_slash(),
      '0'..='9' => self.read_number(),
      'a'..='z' | 'A'..='Z' | '_' => self.read_keyword_or_identifier(),
      '.' => self.read_dot(),
      _ => {
        let mut range = self.create_range();
        let message = format!("Invalid character '{}'", current_char);
        report_and_exit(&message, &mut range, &self.raw, &self.file_name);
      }
    }
  }

  fn read_dot(&mut self) -> Token {
    if self.starts_with("...") {
      self.advance_many(3);
      return Token::new(TokenKind::TripleDot, self.create_range());
    } else {
      self.read_check_ahead("..", TokenKind::Dot, TokenKind::DoubleDot)
    }
  }

  fn read_slash(&mut self) -> Token {
    if self.starts_with("//") {
      self.read_check_ahead("//", TokenKind::DoubleSlash, TokenKind::DoubleSlash)
    } else {
      self.read_check_ahead("/=", TokenKind::Slash, TokenKind::SlashAssign)
    }
  }

  fn read_simple_token(&mut self, kind: TokenKind) -> Token {
    self.advance_one();
    let range = self.create_range();
    Token::new(kind, range)
  }

  fn read_check_ahead(&mut self, expected: &str, single_kind: TokenKind, double_kind: TokenKind) -> Token {
    if self.starts_with(expected) {
      self.advance_many(expected.len());
      let range = self.create_range();
      return Token::new(double_kind, range);
    }
    self.read_simple_token(single_kind)
  }

  fn read_comment_or_minus(&mut self) -> Token {
    if self.starts_with("--") {
      self.read_comment()
    } else {
      self.read_check_ahead("-=", TokenKind::Minus, TokenKind::MinusAssign)
    }
  }

  fn read_comment(&mut self) -> Token {
    if self.starts_with("--[[") {
      self.read_block_comment()
    } else {
      self.read_line_comment()
    }
  }

  fn read_block_comment(&mut self) -> Token {
    self.consume_expect("--[[");
    let text = self.read_until("--]]");
    self.consume_expect_with_custom_error("--]]", "unexpected end of block comment");
    let range = self.create_range();
    Token::new_block_comment(range, text)
  }

  fn read_line_comment(&mut self) -> Token {
    self.consume_expect("--");
    let text = self.read_while(|c| c != '\n');
    let range = self.create_range();
    Token::new_comment(range, text)
  }

  fn read_keyword_or_identifier(&mut self) -> Token {
    let text = self.read_while(|c| c.is_ascii_alphabetic() || c == '_' || c == '$' || c.is_ascii_digit());
    let range = self.create_range();
    Token::new_keyword_or_identifier(range, text)
  }

  fn read_number(&mut self) -> Token {
    let number = self.read_while(match_number);
    let range = self.create_range();
    Token::new_number(range, number)
  }

  fn read_string(&mut self) -> Token {
    self.consume_expect("\"");
    let string = self.read_while(|c| c != '"');
    self.consume_expect_with_custom_error("\"", "unterminated string");
    let range = self.create_range();
    Token::new_string(range, string)
  }

  fn read_single_quote_string(&mut self) -> Token {
    self.consume_expect("'");
    let string = self.read_while(|c| c != '\'');
    self.consume_expect_with_custom_error("'", "unterminated string");
    let range = self.create_range();
    Token::new_string(range, string)
  }

  fn read_while(&mut self, mut test: impl FnMut(char) -> bool) -> String {
    let range_start = self.cursor;
    while !self.is_end() && test(self.peek_one()) {
      self.advance_one();
    }
    self.raw[range_start..self.cursor].to_string()
  }

  fn read_until(&mut self, end: &str) -> String {
    let range_start = self.cursor;
    while !self.is_end() && !self.starts_with(end) {
      let c = self.peek_one();
      if c == '\n' {
        self.advance_new_line();
      } else {
        self.advance_one();
      }
    }
    self.raw[range_start..self.cursor].to_string()
  }

  fn advance_one(&mut self) {
    if let Some(c) = self.raw[self.cursor..].chars().next() {
      self.cursor += c.len_utf8();
      self.column += 1;
      if c == '\n' {
        self.line += 1;
        self.column = 0;
      }
    }
  }

  fn create_range(&mut self) -> Range {
    let start = self.range_start;
    let end = self.cursor;
    self.range_start = self.cursor;
    Range { start, end }
  }

  fn update_current_range(&mut self) {
    self.range_start = self.cursor;
  }

  fn consume_expect(&mut self, text: &str) {
    if self.starts_with(text) {
      self.advance_many(text.len());
    } else {
      let range = self.create_range();
      panic!("expected '{}' at {:?}", text, range);
    }
  }

  fn consume_expect_with_custom_error(&mut self, text: &str, error_message: &str) {
    if self.starts_with(text) {
      self.advance_many(text.len());
    } else {
      let location = self.create_range();
      panic!("{} at {:?}", error_message, location);
    }
  }

  fn is_end(&self) -> bool {
    self.cursor >= self.raw.len()
  }

  fn peek_one(&self) -> char {
    self.raw[self.cursor..].chars().next().unwrap_or('\0')
  }

  fn advance_many(&mut self, count: usize) {
    for _ in 0..count {
      self.advance_one();
    }
  }

  fn starts_with(&self, s: &str) -> bool {
    self.raw[self.cursor..].starts_with(s)
  }

  fn advance_new_line(&mut self) {
    self.line += 1;
    self.column = 0;
    self.cursor += 1;
  }

  fn skip_whitespace(&mut self) {
    while !self.is_end() && self.peek_one().is_whitespace() {
      if self.peek_one() == '\n' {
        self.advance_new_line();
      } else {
        self.advance_one();
      }
    }
  }
}
