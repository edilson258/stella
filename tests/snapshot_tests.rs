#![allow(dead_code, unused_imports)]

use glob::glob;
use insta::assert_ron_snapshot;
use std::ffi::OsString;
use std::fs;
use std::path::Path;
use stella::ast::tokens::{Token, TokenKind};
use stella::lexer::Lexer;
use stella::parser::parser::Parser;

fn read_test_files_with_pattern(pattern: &str) -> Vec<(String, String)> {
  let mut patterns = Vec::new();
  let glob_pattern = glob(pattern).expect("Failed to read glob pattern");
  for entry in glob_pattern {
    let path = entry.expect("Failed to read file");
    let file_name = path.file_name().unwrap().to_string_lossy().to_string();
    let content = fs::read_to_string(path).expect("Failed to read file");
    patterns.push((file_name.to_owned(), content));
  }
  patterns
}

fn create_tokens(source_code: &str) -> Vec<Token> {
  let mut lexer = Lexer::new(source_code.to_string());
  let mut tokens = vec![];
  loop {
    let token = lexer.next_token();
    if token.kind == TokenKind::EOF {
      tokens.push(token);
      break;
    }
    tokens.push(token);
  }
  return tokens;
}

fn create_parser(source_code: &str) -> Parser {
  let parser = Parser::new(source_code);
  return parser;
}
fn format_file_name_with_module(file_name: &str, module: &str) -> String {
  let file_name = format!("{}_{}", module, file_name).replace(".lua", "");
  return file_name;
}
fn setings_snapshot() -> insta::Settings {
  let mut settings = insta::Settings::clone_current();
  settings.set_prepend_module_to_snapshot(false);
  settings.set_omit_expression(true);
  settings
}

#[test]
fn test_lexer_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let tokens = create_tokens(source_code);
      let file_name = format_file_name_with_module(file_name, "lexer");
      assert_ron_snapshot!(file_name.clone(), tokens);
    }
  });
}

#[test]
fn test_parser_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let mut parser = create_parser(source_code);
      let file_name = format_file_name_with_module(file_name, "parser");
      let program = parser.parse_program();
      assert_ron_snapshot!(file_name.clone(), program);
    }
  });
}

#[test]
fn test_type_checker_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  println!("test_type_checker_snapshot: {:?}", test_files);
  // hei, please :) implement me
}
