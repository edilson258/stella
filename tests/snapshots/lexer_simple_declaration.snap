---
source: tests/snapshot_tests.rs
---
[
  Token(
    kind: Local,
    location: Location(
      start: Position(
        line: 1,
        column: 0,
      ),
      end: Position(
        line: 1,
        column: 5,
      ),
    ),
  ),
  Token(
    kind: Identifier("name"),
    location: Location(
      start: Position(
        line: 1,
        column: 6,
      ),
      end: Position(
        line: 1,
        column: 10,
      ),
    ),
  ),
  Token(
    kind: Colon,
    location: Location(
      start: Position(
        line: 1,
        column: 10,
      ),
      end: Position(
        line: 1,
        column: 11,
      ),
    ),
  ),
  Token(
    kind: Identifier("string"),
    location: Location(
      start: Position(
        line: 1,
        column: 12,
      ),
      end: Position(
        line: 1,
        column: 18,
      ),
    ),
  ),
  Token(
    kind: Assign,
    location: Location(
      start: Position(
        line: 1,
        column: 19,
      ),
      end: Position(
        line: 1,
        column: 20,
      ),
    ),
  ),
  Token(
    kind: String("Stella"),
    location: Location(
      start: Position(
        line: 1,
        column: 21,
      ),
      end: Position(
        line: 1,
        column: 29,
      ),
    ),
  ),
  Token(
    kind: EOF,
    location: Location(
      start: Position(
        line: 2,
        column: 0,
      ),
      end: Position(
        line: 2,
        column: 0,
      ),
    ),
  ),
]
